src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -391,19 +394,43 @@ RouterModifyTaskForShardInterval(Query *originalQuery, ShardInterval *shardInter
 	{
 		RelationRestriction *restriction = lfirst(restrictionCell);
 		List *originalBaserestrictInfo = restriction->relOptInfo->baserestrictinfo;
+		Var *relationPartitionKey = PartitionColumn(restriction->relationId,
+													restriction->index);
+		Param *uninstantiatedParameter = NULL;
 
 		/*
-		 * We haven't added the quals if all participating tables are reference
-		 * tables. Thus, now skip instantiating them.
+		 * We don't need to add restriction to reference tables given that they are
+		 * already single sharded and always prune to that single shard.
 		 */
-		if (allReferenceTables)
+		if (!relationPartitionKey)
```

Comment:
I'd slightly prefer testing for this actually being a reference table, rather than nullness of partition column. Seems cleaner.